README.md
# flog-experiments

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2016 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-11-01
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-11-01
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/flog-experiments/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/flog-experiments/compare/0.1.0...0.1.1


test/flog_experiments/core_test.clj
(ns flog-experiments.core-test
  (:require [clojure.test :refer :all]
            [flog-experiments.core :refer :all]))


test/flog_experiments/test_util.clj
(ns flog-experiments.test-util
  (:require [flog-experiments.wiremock :as wm]))

(defn wiremock [f]
  (let [s (wm/server)]
    (wm/start s)
    (f)
    (wm/stop s)))


test/flog_experiments/trad_test.clj
(ns flog-experiments.trad-test
  (:require [clojure.test :refer :all]
            [flog-experiments.servers :as servers]
            [flog-experiments.trad :as trad]))

(deftest no-errors-server
  (is (= ["/img/clojure.jpg"
          "/img/logo.jpg"
          "/img/mainpage.jpg"
          "/img/me.jpg"
          "/img/rock-climbing.jpg"]
         (servers/with-server servers/no-errors
           #(trad/crawled-images "http://localhost:8080" "/index.html")))))

(deftest errors-server
  (is (= ["/img/logo.jpg"
          "/img/mainpage.jpg"
          "/img/me.jpg"]
         (servers/with-server servers/errors
           #(trad/crawled-images-error-handling "http://localhost:8080" "/index.html")))))

(deftest timeouts-server
  (is (= ["/img/logo.jpg"
          "/img/mainpage.jpg"
          "/img/me.jpg"]
         (servers/with-server servers/timeouts
           #(trad/crawled-images-error-handling "http://localhost:8080" "/index.html")))))


src/flog_experiments/trad.clj
(ns flog-experiments.trad
  (:require [flog-experiments.servers :as servers]
            [clj-http.client :as http]
            [hickory.core :as h]
            [clojure.string :as string]
            [hickory.select :as h.select]))

(defn full-url [domain path]
  (if (string/starts-with? path "/")
    (str domain path)
    path))

(defn internal-urls [domain body]
  (->> body
       h/parse
       h/as-hickory
       (h.select/select (h.select/tag :a))
       (map (comp (partial full-url domain) :href :attrs))))

;; Note you can already test whether only 
;; internal domains are selected here.
(defn internal-images [body]
  (->> body
       h/parse
       h/as-hickory
       (h.select/select (h.select/tag :img))
       (map (comp :src :attrs))))

(defn temporary-error? [x]
  (#{503 :timeout} x))

(defn non-fatal-error? [x]
  (#{404} x))

(defn success? [x]
  (#{200 201 204} x))

(defn safe-get [url]
  (try
    (http/get url {:throw-exceptions false
                   :socket-timeout 500
                   :conn-timeout 500})
    (catch java.net.SocketTimeoutException e
      {:status :timeout})))


(defn crawled-images-error-handling
  "Crawls URL and returns a list of images found within that domain.
  
  Handles the following errors:
  * if response is 404, skips page
  * if response is 503, sleeps and tries again (3 times and then skips)
  * if response times out, sleeps and tries again (3 times and then skips)"
  [domain url]
  (loop [visited-urls #{}
         [u & remaining-urls :as urls] [(full-url domain url)]
         images []
         errors {}]
    (if (nil? u)
      (sort images)
      (let [response (safe-get u)
            {:keys [body status]} response]
        (cond
          (success? status)
          (let [new-imgs (internal-images body)
                new-urls (remove visited-urls (internal-urls domain body))]
            (recur (into visited-urls [u])
                   (into remaining-urls new-urls)
                   (distinct (into images new-imgs))
                   errors))
          
          (non-fatal-error? status)
          (recur (into visited-urls [u])
                 remaining-urls
                 images
                 errors)

          (temporary-error? status)
          (if (< (count (get errors u)) 2)
            (do
              ;; wait
              (Thread/sleep 50)
              (recur visited-urls
                     urls
                     images
                     (update errors u conj status)))
            (recur (into visited-urls [u])
                   remaining-urls
                   images
                   errors))
          
          :else
          (throw (ex-info "Server returned unrecoverable response" {:response response})))))))

;; public ;;;;

(defn crawled-images
  "Crawls URL and returns a list of images found within that domain."
  [domain url]
  (loop [visited-urls #{}
         [u & remaining-urls] [(full-url domain url)]
         images []]
    (if u
      (let [body (:body (http/get u))
            new-imgs (internal-images body)
            new-urls (remove visited-urls (internal-urls domain body))]
        (recur (into visited-urls [u])
               (into remaining-urls new-urls)
               (distinct (into images new-imgs))))
      (sort images))))

(comment
  (servers/with-server
    servers/errors
    #(crawled-images-error-handling "http://localhost:8080" "/index.html"))
  )


src/flog_experiments/wiremock.clj
;; take from https://github.com/alexanderjamesking/clj-wiremock
;; so I can tweak it as needed
(ns flog-experiments.wiremock
  (:require [cheshire.core :refer [generate-string parse-string]]
            [clj-http.client :as client])
  (:import com.github.tomakehurst.wiremock.WireMockServer
           com.github.tomakehurst.wiremock.core.WireMockConfiguration))

(defn config
    "Creates a new instance of WireMockConfiguration"
  ([] (new WireMockConfiguration))
  ([config-map]
   (let [config (new WireMockConfiguration)]
     (when-let [v (:port config-map)] (.port config (int v)))
     (when-let [v (:https-port config-map)] (.httpsPort config (int v)))
     (when-let [v (:keystore-path config-map)] (.keystorePath config v))
     (when-let [v (:keystore-password config-map)] (.keystorePassword config v))
     (when-let [v (:trust-store-path config-map)] (.trustStorePath config v))
     (when-let [v (:trust-store-password config-map)] (.trustStorePassword config v))
     (when-let [v (:need-client-auth config-map)] (.needClientAuth config v))
     config)))

(defn server
    "Create a new instance of WireMockServer"
  ([] (new WireMockServer))
  ([config] (new WireMockServer config)))

(defn start
    "Starts the WireMockServer"
  [server]
  (.start server))

(defn stop
    "Stops the WireMockServer"
  [server]
  (.stop server))

(defn reset
    "Removes all stub mappings and deletes the request log"
  [server]
  (.resetMappings server))

(defn- admin-post [endpoint body & [base-url]]
  (let [base-url (or base-url "http://localhost:8080")
        admin-url (str base-url "/__admin/" endpoint)
        content-to-post {:body (generate-string body)}
        response (client/post admin-url content-to-post)]
    (parse-string (:body response) true)))

(defn count-requests [body & [base-url]]
  (:count (admin-post "requests/count" body base-url)))

(defn find-requests [body & [base-url]]
  (:requests (admin-post "requests/find" body base-url)))

(defn stub [body & [base-url]]
  (admin-post "mappings/new" body base-url))

(comment
  (def s (server))
  (let [s (server)]
    (start s)
    ;; check http://localhost:8080/__admin/ for config
    (stub {:request {:method "GET" :url "/hello"}
           :response {:status 200 :body "Hello World"}})
    (client/get "http://localhost:8080/hello")
    (stop s)
    )
  )


src/flog_experiments/servers.clj
(ns flog-experiments.servers
    (:require [flog-experiments.wiremock :as wm]
              [hiccup.page :as h.page]
              [hiccup.element :as h.el]))

(def index-page
  (h.page/html5
   [:body
    [:h1
     "My page"
     ]
    (h.el/image "/img/logo.jpg")
    (h.el/image "/img/mainpage.jpg")
    (h.el/unordered-list
     [(h.el/link-to "/index.html" "Home")
      (h.el/link-to "/about.html" "About")
      (h.el/link-to "/interests.html" "Interests")])]))

(def about-page
  (h.page/html5
   [:body
    [:h1
     "About"]
        (h.el/image "/img/logo.jpg")
    (h.el/image "/img/me.jpg")
    (h.el/unordered-list
     [(h.el/link-to "/index.html" "Home")
      (h.el/link-to "/interests.html" "Interests")])]))

(def interests-page
  (h.page/html5
   [:body
    [:h1
     "Interests"]
    (h.el/image "/img/logo.jpg")
    (h.el/image "/img/clojure.jpg")
    (h.el/image "/img/rock-climbing.jpg")
    (h.el/unordered-list
     [(h.el/link-to "/index.html" "Home")
      (h.el/link-to "/about.html" "About")])]))

(defn no-errors []
  (doseq [sc [{:request {:method "GET" :url "/index.html"}
               :response {:status 200 :body index-page}}
              {:request {:method "GET" :url "/about.html"}
               :response {:status 200 :body about-page}}
              {:request {:method "GET" :url "/interests.html"}
               :response {:status 200 :body interests-page}}
              ]]
    (wm/stub sc)))

(defn errors []
  (doseq [sc [{:scenarioName "Server overloaded"
                :requiredScenarioState "Started"
                :newScenarioState "Some overloaded"
                :request {:method "GET" :url "/index.html"}
               :response {:status 503}}
              {:scenarioName "Server overloaded"
                :requiredScenarioState "Started"
                :request {:method "GET" :url "/about.html"}
               :response {:status 503}}
              {:scenarioName "Server overloaded"
                :requiredScenarioState "Started"
                :request {:method "GET" :url "/interests.html"}
               :response {:status 503}}
              ;;;;;;;;;;;;;;;;;;;
              {:scenarioName "Server overloaded"
               :requiredScenarioState "Some overloaded"
               :newScenarioState "All OK"
               :request {:method "GET" :url "/index.html"}
               :response {:status 200 :body index-page}}
              {:scenarioName "Server overloaded"
               :requiredScenarioState "Some overloaded"
               :newScenarioState "All OK"
               :request {:method "GET" :url "/about.html"}
               :response {:status 503}}
              {:scenarioName "Server overloaded"
               :requiredScenarioState "Some overloaded"
               :newScenarioState "All OK"
               :request {:method "GET" :url "/interests.html"}
               :response {:status 503}}
              ;;;;;;;;;;;;;;;;;;;
              {:scenarioName "Server overloaded"
               :requiredScenarioState "All OK"
               :request {:method "GET" :url "/index.html"}
               :response {:status 200 :body index-page}}
              {:scenarioName "Server overloaded"
               :requiredScenarioState "All OK"
               :request {:method "GET" :url "/about.html"}
               :response {:status 200 :body about-page}}
              {:scenarioName "Server overloaded"
               :requiredScenarioState "All OK"
               :request {:method "GET" :url "/interests.html"}
               :response {:status 404}}]]
    (wm/stub sc)))

(defn timeouts []
  (doseq [sc [{:request {:method "GET" :url "/index.html"}
               :response {:status 200
                          :body index-page
                          :fixedDelayMilliseconds 20}}
              {:request {:method "GET" :url "/about.html"}
               :response {:status 200
                          :body about-page
                          :fixedDelayMilliseconds 200}}
              {:request {:method "GET" :url "/interests.html"}
               :response {:status 200
                          :body interests-page
                          :fixedDelayMilliseconds 600}}]]
    (wm/stub sc)))

(defn with-server [setup f]
  (let [s (wm/server)]
    (wm/stop s)
    (wm/start s)
    (try
      (setup)
      (f)
      (finally
        (wm/stop s)))))
